aten/src/ATen/native/SharedReduceOps.h

```diff
@@ -284,16 +284,17 @@ struct NormOneOps {
   }
 
 #if defined(__CUDACC__) || defined(__HIPCC__)
-  inline C10_DEVICE acc_t warp_shfl_down(acc_t data, int offset) const {
-    return WARP_SHFL_DOWN(data, offset);
+  inline C10_DEVICE acc_t warp_shfl_down(acc_t acc, int offset) const {
+    return WARP_SHFL_DOWN(acc, offset);
   }
 #endif
 };
 
-template <typename acc_t>
+template <typename scalar_t, typename acc_t=scalar_t>
 struct NormTwoOps {
-  inline C10_DEVICE acc_t reduce(acc_t acc, acc_t data, int64_t /*idx*/) const {
-    return acc + data * data;
+  inline C10_DEVICE acc_t reduce(acc_t acc, scalar_t data, int64_t /*idx*/) const {
+    acc_t abs_data = std::abs(data);
```

Comment:
What's the difference between the previous expression and the new one with `std::abs`?

Reply:
The `std::abs()` call is needed when `scalar_t` is complex. For real numbers, it's not needed. I could add an overload for complex numbers so that we avoid calling `std::abs()` for real numbers.

Reply:
Adding an overload sounds good as long as it doesn't add too much code complexity.

Reply:
I added `abs_if_complex()`. Not too sure if this is the most efficient way to implement it. Let me know if you think this adds too much complexity.